### DISCUSSION TITLE
crazyflie cannot hover with two V1 base stations and a lighthouse deck

### DISCUSSION THREAD

**Discussion by seunghyunnn** [bitcraze/discussions#118] [Q&A]
H, I'm using HTC Vive base stations (Lighthouse V1) and a lighthouse deck for positioning the drone. Currently, I am flying my crazyflie charged more than 4, in the square-shaped place using two base stations facing each other at each end of a diagonal (they are not connected with each other). I use a python-based cfclient GUI to set up, check position and fly the drone. 
The issues I have : 
1) The 'Flight Control' tab shows that the drone is keep shaking even though the drone is on the floor
2) If I fly the drone by pressing the 'Take off' button in the 'Flight Control' tab it moves in random places. (It cannot hover. It moves to the end of the room, to the corner of the room. It flips and fall.)  

Then to fix these issues, I tried two things: 
1) turning off one of the Lighthouse stations 
2) and as the z-axis in the 'Lighthouse Positioning' tab was when I put the drone on the floor, I set the drone on the box to have a positive z-axis position 
With both of the methods applied,  the 'Take off' was comparatively stable (it was still wiggling sometimes, but it was somehow able to stay comparatively still, hovering on one place)

What would be the reason for the crazyflie going random places and cannot hover under these circumstances? Are there any suggestions to fix these errors?  
(I want to use 2 lighthouse stations as I need the precise positioning, and I want to be able to start my drone on the floor. )

Thank you for your time and support.

**Comment by krichardsson** [comment]
Hi!

OK, let's try to figure this out.

First some comments and explanations of what you see

> The 'Flight Control' tab shows that the drone is keep shaking even though the drone is on the floor

When the Crazyflie receives a light sweep from one of the base stations, it will use the position of the base station that transmitted the sweep (stored in the system geometry) to calculate where the light plane is located in the global coordinate system. The information is then pushed to the kalman state estimator to update the estimated state (position and orientation) of the Crazyflie. In a system where everything is OK, each update will usually just be a small adjustment of the current state, maybe because the Crazyflie has move a little bit. On the other hand, if something is not correct in the system, the information that is pushed to the estimator will be very different from the current state and the update will cause change to the state. If the sweeps from the two base stations provide different information to the estimator the result will be something like what you describe. The position might drift away and/or the orientation may be "going bananas".

> If I fly the drone by pressing the 'Take off' button in the 'Flight Control' tab it moves in random places. (It cannot hover. It moves to the end of the room, to the corner of the room. It flips and fall.)

If the system does not work properly and you have a stable position when the Crazyflie is on the floor or moved around in your hand, it will not be able to fly either. A stable position estimation is essential for flight.

> turning off one of the Lighthouse stations

When only one base station is on, the result is that there is no conflicting information and the position should be stabilize. If the position of the base station in the system geometry is wrong, the estimated position will not be correct though. 

I think there are two things to check first:
1. Check that the synchronization between the base stations works. As you are not using a synch cable, the base stations should run on modes "b" and "c". If "receiving" and "calibration" are green for both base stations in the "Base station status" box, in the "Lighthouse Position" tab of the client, all should be good.
2. The system geometry. If the geometry, that is the position and orientation of the base stations, is not correct, the Crazyflie will get conflicting information from the two base stations. There are two ways of calculating the system geometry: the first is by using client as described in [the getting started guide](https://www.bitcraze.io/documentation/tutorials/getting-started-with-lighthouse/#preparing-the-system), the other option is to use the new functionality in the python lib, [documented here, step 5 should be all you need](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/lighthouse/multi_base_stations/). I strongly recommend to use the second option (the script in the python lib) as it will provide a much better result. What ever way you chose, check that the coordinates you get for the base stations looks reasonable, if not there might be a problem when measuring/calculating the system geometry.

My guess is that you need to re-estimate your system geometry, but let us know what you find out and we will continue from there.

**Reply by seunghyunnn** [reply]
Thank you for the reply! It gave me some useful insights to understand the lighthouse system more advanced and possible sources that the issue is coming from. As you said, there might be an issue in the geometry; two stations may give different geometry which confuses the drone. 

1. Regarding the first point, I checked the synchronization between the base station and I guess it is working fine. I am not using the synching cable, so I ran the mode 'b' and 'c'. Every box was green in the "Lighthouse Position" tab. 
2. I guess the issue is coming from the second point that you have mentioned. I executed the code that you have suggested, multi_bs_geometry_estimation.py and I got the following result. 

/////////////////////////////
Step 6. Estimating geometry...
/Users/seunghyun/Desktop/crazyflie-lib-python/cflib/localization/lighthouse_types.py:61: ComplexWarning: Casting complex values to real discards the imaginary part
  return Pose(Rotation.from_quat(R_quat).as_matrix(), t_vec)
Initial guess base stations at:
    1: (-1.344640072356544, -1.6813715371521656, 1.0380659921475308)
    2: (-1.3008794858447432, -0.7358374828625579, 1.093352815897955)
17 samples will be used
Solution did not converge, it might not be good!
Raw solution:
  Base stations at:
    1: (-1.4083649392854285, -0.7377892801274163, 1.1128004026521698)
    2: (-1.374119237048485, -0.7138195222431858, 1.0943544446962488)
  Solution match per base station:
    1: {'mean_error': 0.008932659057995488, 'max_error': 0.02424333063160657, 'std_error': 0.00495639891345788}
    2: {'mean_error': 0.007679848490502319, 'max_error': 0.01991167929201481, 'std_error': 0.004612205545871596}

Final solution:
  Base stations at:
    1: (1.8895201047172714, -0.5429949179394375, 0.020415610484377664)
    2: (1.847681173168706, -0.5255425090131454, 0.011057000601317635)
  Geometry estimated
Step 7. Upload geometry to the Crazyflie
Press enter to upload geometry. 
Geometry uploaded
///////////////////////////////////

  As you can see, it says 'Solution did not converge, it might not be good!' which means that I didn't get the acceptable solution from the 'solve' method (https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/localization/lighthouse_geometry_solver/ ). 
  
  Is there anything I can do to make the 'solve' method find the 'acceptable solution'? 
  
  Do you think I should update the firmware of the V1 stations that I am using? (https://business.vive.com/us/support/vive-pro-secure/category_howto/updating-the-base-stations-via-cable-connection.html) But it was working fine previously -about an year ago- and I am afraid if I would mess up the versions. 
  
  Or do you think using the cable would solve this issue? 
  
  I am also open to any other possible solutions!
  
  Thank you so much for your time and support:)

**Reply by krichardsson** [reply]
Great!

If the base stations look happy and you see that you receive both base stations in the client, I think you can assume the synchronization works. I don't think a cable would make a difference.
I also agree that you should not upgrade the base stations, if they worked a year ago the firmware should be good.

So we are left with the geometry, let's see if we can figure out why that does not work.
The geometry estimation script is processing the data in several steps:
1. Make a rough estimate using the IPPE method. IPPE is fast but returns two possible solutions for each recorded position and the script tries to figure out which of the solutions that is the correct one by matching solutions from the recorded positions. The result of this process is the "initial guess" in your log. You mentioned that your base stations were placed in opposing corners of your flight space, but according to the initial guess they are located almost in the same point! It seems as something went wrong at this step already.
2. Improve the initial guess by solving a large equation system. The output from this step is logged as "Raw solution"
3. Scaling and rotating the system based on the positions of the Crazyflie in the first steps of the collection process. This is logged as the "Final solution"

As you noted the solution did not converge, which also makes sense based on the poor initial guess.
A note on the coordinate system used in the logging of step 1 and 2: the coordinate system is based on the pose of the Crazyflie in the first recorded position, so if it was not oriented facing positive X the coordinates will be rotated. 

Now, the question is why the initial guess fails?
I have 3 things you could check
1. Reflections can cause problems as the Crazyflie can not distinguish reflections from direct light from a base station which might trick the Crazyflie to receive the wrong sweep angles. Do you have windows, mirrors, metal surfaces in the room that might cause problems?
2. You could try to move the base stations a bit to see if it makes any difference. It is possible that the geometry estimation script might pick the correct solution if the setup is a bit different.
3. It is possible to log the angles received by the lighthouse deck to examine if they look OK when you move the Crazyflie around a bit. [Set up the plotter](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/plotter_tab/) to show logging variables lighthouse.angle0x and lighthouse.angle0y (from the first base station) as well as lighthouse.angle1x and lighthouse.angle1y (from the second base station). If you move the Crazyflie sideways (in relation to the base station) the x-angle should change and if you move it up and down the y-angle should change. The change should be nice and smooth without jumps.

If you have any HTC Vive gaming equipment (like the hand controllers or the VR headset), you could try the system with those to ensure the base stations are fine and synchronized.

**Reply by seunghyunnn** [reply]
Thank you for your reply! 

Firstly, I have checked the 3 things that you said. 
1. There was a window in my space, so I blocked every sunlight. I even tried covering the floor with papers so that it couldn't reflect anything but it didn't work. 
2. I moved the base stations a little bit, and also tried positioning them in the opposite direction of the room but the wiggle continued. 
3. Thank you for the guidance to the plotter! I set up the plotter and the x-angle and y-angle were smoothly changing. 

While I was following your suggestion, I found two strange points that may affect my crazyflie to wiggle. 
1. I played around with the plotter and draw the plot of stateEstimate.x, stateEstimate.y, stateEstimate.z, StateEstimate.pitch, StateEstimate.roll, and StateEstimate.yaw, and everything that I listed here was drawing the wiggling graph. 
<img width="1384" alt="Screen Shot 2022-07-01 at 1 14 45 PM" src="https://user-images.githubusercontent.com/67893627/177076606-e5a1acbb-a157-46b9-8576-adf3061f27d5.png">

2. Moreover, when I look at the console, I found out that the console is repeatedly logging 'ESTKALMAN: State out of bounds, resetting'. 

Do you have any idea what may cause these issues? I searched for the 'ESTKALMAN : State out of bounds, resetting' issue in the bitcraze forum, but I couldn't get a relevant solution. 

Moreover, do you think PID tuning will make any difference?  (I searched the bitcraze forum and some people were suggesting PID tuning for wobbling issue) 


Just in case, the full console was : 
SYS: Crazyflie 2.1 is up and running!
SYS: Production release 2021.06
SYS: I am 0x2030334B554B50190055002B and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_INFO: Warning! No driver found for deck.
DECK_CORE: 2 deck(s) found
DECK_CORE: Calling INIT on driver bcLighthouse4 for deck 1
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 19248 bytes
LHFL: Lighthouse bootloader version: 2
STAB: Ready to fly.
LHFL: Bitstream CRC32: 112BC794 [PASS]
LHFL: Firmware version 6 verified, booting deck!
ESTKALMAN: State out of bounds, resetting
ESTKALMAN: State out of bounds, resetting
ESTKALMAN: State out of bounds, resetting


Thank you so much:)

**Comment by krichardsson** [comment]
Thanks for the info!

> Do you have any idea what may cause these issues? I searched for the 'ESTKALMAN : State out of bounds, resetting' issue

This message is displayed when the kalman supervisor finds that the state in the kalman filter is unreasonable, for instance if it is representing a very high velocity or a position very far away. It is an indication that the estimator can not make sense of the sensor data (lighthouse information as well as gyros, acceleratometers or any other connected sensor) and does not converge to a stable solution.

Did you also plot the logging variables lighthouse.angle0x, lighthouse.angle0y, lighthouse.angle1x and lighthouse.angle1y?

> do you think PID tuning will make any difference? 

No, that will not help. The PID controller is using the data from the estimator, and if the estimator does not have a good understanding of the state, the controller can not do anything at all.

> DECK_INFO: Warning! No driver found for deck.
DECK_CORE: 2 deck(s) found

I can see in the log that you have a second deck attached to the Crazyflie but that it has not been identified. What deck is that?

**Reply by seunghyunnn** [reply]
Thank you for the reply:) 
1. 
<img width="705" alt="Screen Shot 2022-07-06 at 10 34 01 AM" src="https://user-images.githubusercontent.com/67893627/177448792-83786407-8769-4bbc-83ff-a7f5110f6ad5.png">
For the lighthouse angle, they have little wiggles but are fairly smooth. There are some uneven signs like 98-99 in the above image but they weren’t extremely uneven. 
2.
The two decks were lighthouse deck and the wireless charging deck.  
3. Additionally, at the console, I sometimes get ' ESTKALMAN: WARNING: Kalman prediction rate low (98)' and sometimes get 'ESTKALMAN: State out of bounds, resetting'

Is there any way to check the gyros, accelerometers or any other connected sensors individually? and also how can I make them converge?

**Reply by tobbeanton** [reply]
> Is there any way to check the gyros, accelerometers or any other connected sensors individually?

You can disconnect any decks and connect using the cfclient and look at the virtual horizon. Does this behave correctly? If you have the possibility to manually fly it (joystick) that would also be a good test.

**Reply by krichardsson** [reply]
The lighthouse angles look good. The flat parts are due to that one of the base stations was not received for a while.

> Is there any way to check the gyros, accelerometers or any other connected sensors individually? 

You can plot them the same way as the lighthouse angles, you will find them in the `acc` and `gyro` grouops. Since you can fly with one base station, they are probably fine.

> and also how can I make them converge?

I still think the problem is that the base station geometry estimation script does not find the correct position of your base stations and I think this is where you should put your efforts. I'm not sure why it is not working, but I would probably try to move the base stations around a bit or move/rotate your coordinate system a bit to see if you can make the script find a proper solution.
